### PR TITLE
- Apply transaction before any service call

### PR DESCRIPTION
**Pull request by alaibe**
- Create a service job to reduce boilerplate code

Pls tell me if you see any problems in transaction (about mysql ock or thread lock, concurrency...)

### CODE REVIEW

**Comment by kurbaitis** [comment]
Pdf generating takes time, so not sure if deadlocks could appear.

**Comment by alaibe** [comment]
PDF generation should go in background

**Comment by kurbaitis** [comment]
Then that PDF jobs should be migrated... We work with objects as arguments for service call, still question: is it working good with all services?

**Comment by alaibe** [comment]
No if you call: call_later you need to pass something serializable:
http://edgeapi.rubyonrails.org/classes/ActiveJob/SerializationError.html

The call_later is to avoid boilerplate call see:

`CustomerShortcutCreateJob.perform_later(customer.id)` then
```
class CustomerShortcutCreateJob < ActiveJob::Base
   def perform(customer_id)
      Spanner::CustomerShortcuts::Create.call customer_id: customer_id
   end
end
```

to:
`Spanner::CustomerShortcuts::Create.call_later customer_id: customer.id`

**Comment by alaibe** [comment]
BTW you can pass object that are saved (have an id):
Spanner::CustomerShortcuts::Create.call_later customer: customer
=> that still work

**Comment by os6sense** [comment]
👎 

- transactions only rollback on exceptions; anywhere in the code we aren't using bang methods (e.g. update_attributes) there is the possibility of the transaction not being rolled back
- This is going to create nested transactions; its been a while since I used these but I recall some funky-ness being required for child transactions. http://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html
- Callbacks might not work as expected after adding a transactional wrapper.

Basically I would expect a lot of unexpected consequences from introducing this into a service (including that a service by itself is not necessarily transactional)

**Comment by alaibe** [comment]
1 we always use ! method
2 Nested transaction seems to be usefull when you use mutliple data source
3 We don't use callbacks :)

**Comment by alaibe** [comment]
About the case by case basis, do you know how many do we have already: 0 so no one think about it.
And it is quite impossible to manage actually

**Comment by os6sense** [comment]
By adding a (e.g) transactional: false parameter to specify where the operation is NOT transactional this will allow us a bit more granularity and control of how the service behaves and allow us to see which services do ont touch the db.
